backend/scheduler.js

```diff
@@ -0,0 +1,204 @@
+                break;
+            }
+        }
+
+        if (nextEvent) {
+            profitGoing = eventList[i].profit + maxProfit[nextEvent];
+        } else {
+            profitGoing = eventList[i].profit;
+        }
+
+        // Not going. We add the next event's max profit, if available
+        let profitSkip = 0;
+        if (i < eventList.length - 1) {
+            profitSkip = maxProfit[i + 1];
+        }
+
+        // Compare both profits, choose higher
+        if (profitGoing > profitSkip) {
+            maxProfit[i] = profitGoing;
+            goOrSkip[i] = nextEvent;
+        } else {
+            maxProfit[i] = profitSkip;
+            goOrSkip[i] = "SKIP";
+        }
+    }
+
+    const result = [];
+    let i = 0;
+    while (i !== null && i < eventList.length) {
+        if (goOrSkip[i] !== "SKIP") {
+            result.push(eventList[i].id);
+            i = goOrSkip[i];
+        } else {
+            i++;
+        }
+    }
+
+    return { selectedEventIds: result, totalProfit: maxProfit[0] };
+}
+
+/**
+ * Returns commute penalty for a commute tie in minutes
+ * @param {number} commuteTime in minutes
+ * @returns float between 0 and 1
+ */
+function commutePenalty(commuteTime) {
+    return Math.min(1, 2.1718 ** (-0.02 * (commuteTime - 30)));
+}
+
+/**
+ * Given events with start time and end time, as well as a 'profit'. Maximize profit without overlapping events.
+ * Take into account commute time between events. (driving)
+ * @param {object[]} events array of events. must have start_time, end_time and profit. Must have location (lng, lat)
+ * @returns list of event ids to go to and the profit
+ */
+async function scheduleWithCommutes(events) {
+    const eventList = events.toSorted((a, b) => {
+        return new Date(a.start_time) - new Date(b.start_time);
+    });
+
+    // Initialize dp arrays
+    const maxProfitMemo = [];
+    const goOrSkip = [];
+    for (let _ of eventList) {
+        maxProfitMemo.push(null);
+        goOrSkip.push(null);
+    }
+
+    const commuteMemo = new Map()
+
+    const maxProfit = async (i) => {
+        if (i >= eventList.length) {
+            return 0;
+        }
+        if (maxProfit[i]) {
+            return maxProfit[i];
+        }
+
+        // Going, cannot only check first event that fits, will check every event and add commute penalty
+        const nextProfits = [];
+        for (let j = i + 1; j < eventList.length; j++) {
+            // Create a string key to memoize commute results
+            const key = JSON.stringify([eventList[i].id, eventList[j].id].sort((a,b) => {return a - b}))
+            const commuteTime = commuteMemo.has(key) ? commuteMemo.get(key) : await calculateCommute(
+                eventList[i],
+                eventList[j]
+            );
+            commuteMemo.set(key, commuteTime)
+            if (eventList[j].start_time > eventList[i].end_time + commuteTime) {
+                nextProfits.push(
+                    [j,
```

Comment:
I think you could simplify this part for readability and to save some storage. It looks like you're just trying to track the max profit in this loop, so instead of storing everything in an array and looping through that array later, you can just search for the max in this loop. I would also recommend having separate vars for "j" and the maxProfit, rather than combining them in an array.